### PR TITLE
Three progresses from linkai

### PR DESCRIPTION
**Pull request by songlkkevin**
1. fix a probable out of bound access.
2. Implement the scaling up part.
3. fix the memory leakage bug

### CODE REVIEW

**Comment by pratiksampat** [comment]
Patch 1 and 3 LGTM.

File: kernel/sched/fair.c [hunk 1]
```diff
@@ -6174,9 +6172,11 @@ static int do_sched_cfs_period_timer(struct cfs_bandwidth *cfs_b, int overrun, u
 
 		if (cfs_b->curr_interval < 10) {
 			cfs_b->recommender_active = false;
+			if (throttled && cfs_b->curr_interval > 1)
+				cfs_b->max_ratio = min((unsigned long long)num_online_cpus() * 100000, cfs_b->max_ratio * 2);
```
Comment (pratiksampat) [inline kernel/sched/fair.c:6176]: do we need to reset max_ratio after the scale up is done and throttle is mitigated?

File: kernel/sched/sched.h [hunk 2]
```diff
@@ -388,6 +388,7 @@ struct cfs_bandwidth {
 	u64			*pb_period_hist;
 	u64			*pb_runtime_hist;
 	u64			pb_millicpu;
+	u64         max_ratio;
```
Comment (pratiksampat) [inline kernel/sched/sched.h:391]: nit: indent

File: kernel/sched/fair.c (hunk 1 above)
Comment (songlkkevin) [reply kernel/sched/fair.c:6176]: I intended to do this. In this case, when next burst happens, it can satisfy the burst quickly.

**Pull request merged**